sw/device/sca/aes_serial/simple_serial.c

```diff
@@ -0,0 +1,218 @@
+ * @param cmd A simple serial command.
+ * @return Command handler's index in `handlers`.
+ */
+static size_t simple_serial_get_handler_index(uint8_t cmd) {
+  if (simple_serial_is_valid_command(cmd)) {
+    return cmd - 'a';
+  } else {
+    return ARRAYSIZE(handlers) - 1;
+  }
+}
+
+/**
+ * Simple serial 'v' (version) command handler.
+ *
+ * Returns the simple serial version that this file implements. This command is
+ * useful for checking that the host and the device can communicate properly
+ * before starting capturing traces.
+ *
+ * @param data Received packet payload.
+ * @param data_len Payload length.
+ */
+static void simple_serial_version(const uint8_t *data, size_t data_len) {
+  simple_serial_send_status(kSimpleSerialProtocolVersion);
+}
+
+/**
+ * Simple serial 's' (seed PRNG) command handler.
+ *
+ * This function only supports 4-byte seeds.
+ *
+ * @param seed A buffer holding the seed.
+ * @param seed_len Seed length.
+ */
+static void simple_serial_seed_prng(const uint8_t *seed, size_t seed_len) {
+  SS_CHECK(seed_len == sizeof(uint32_t));
+  prng_seed(read_32(seed));
+}
+
+/**
+ * Handler for uninmplemented simple serial commands.
+ *
+ * Sends an error packet over UART.
+ *
+ * @param data Received packet payload
+ * @param data_len Payload length.
+ */
+static void simple_serial_unknown_command(const uint8_t *data,
+                                          size_t data_len) {
+  simple_serial_send_status(kSimpleSerialError);
+}
+
+void simple_serial_init(const dif_uart_t *uart_) {
+  uart = uart_;
+
+  for (size_t i = 0; i < ARRAYSIZE(handlers); ++i) {
+    handlers[i] = simple_serial_unknown_command;
+  }
+  handlers[simple_serial_get_handler_index('s')] = simple_serial_seed_prng;
+  handlers[simple_serial_get_handler_index('v')] = simple_serial_version;
+}
+
+simple_serial_result_t simple_serial_register_handler(
+    uint8_t cmd, simple_serial_command_handler handler) {
+  if (!simple_serial_is_valid_command(cmd)) {
+    return kSimpleSerialError;
+  } else if (cmd == 's' || cmd == 'v') {
+    // Cannot register handlers for built-in commands.
+    return kSimpleSerialError;
+  } else {
+    handlers[simple_serial_get_handler_index(cmd)] = handler;
+    return kSimpleSerialOk;
+  }
+}
+
+void simple_serial_process_packet(void) {
+  uint8_t cmd;
+  uint8_t data[kUartMaxRxPacketSize];
+  size_t data_len;
+  simple_serial_receive_packet(&cmd, data, ARRAYSIZE(data), &data_len);
+  handlers[simple_serial_get_handler_index(cmd)](data, data_len);
+}
+
+void simple_serial_send_packet(const uint8_t cmd, const uint8_t *data,
+                               size_t data_len) {
+  base_printf("%c", cmd);
+  simple_serial_print_hex(data, data_len);
+  base_printf("\n");
+}
+
+void simple_serial_send_status(uint8_t res) {
+  simple_serial_send_packet('z', (uint8_t[1]){res}, 1);
+}
+
+void simple_serial_print_hex(const uint8_t *data, size_t data_len) {
+  for (int i = 0; i < data_len; ++i) {
+    base_printf("%2x", (uint32_t)data[i]);
```

Comment:
nit: this doesn't look safe. Does this work still with `aes_serial_single_encrypt`?  Should we remove this API?

Reply:
I guess you referring to the uint32_t cast, right? It shouldn't be there, thanks for cathching this! I will fix this in the next update.

I exposed this function because I thought it could be useful to print buffers when developing or debugging.